packages/okta-adapter/src/client/client.ts

```diff
@@ -58,24 +136,45 @@ export default class OktaClient extends clientUtils.AdapterHTTPClient<
       {
         pageSize: DEFAULT_PAGE_SIZE,
         rateLimit: DEFAULT_MAX_CONCURRENT_API_REQUESTS,
-        // TODO SALTO-2649: add better handling for rate limits
         maxRequestsPerMinute: DEFAULT_MAX_REQUESTS_PER_MINUTE,
-        // wait for 10s before trying again, change after SALTO-2649
-        retry: { ...DEFAULT_RETRY_OPTS, retryDelay: 10000 },
+        retry: DEFAULT_RETRY_OPTS,
       }
     )
+    this.rateLimitBuffer = clientOpts.rateLimitBuffer ?? DEFAULT_RATE_LIMIT_BUFFER
   }
 
   public get baseUrl(): string {
     return this.credentials.baseUrl
   }
 
+  // This getSinglePage tracks the number of running requests per endpoint, with their rate limits and reset times
+  // It Pauses new requests when running requests approach the rate limit, factoring in a buffer
+  // Then it resumes when the rate limit resets.
+  // For more information: SALTO-4350
   public async getSinglePage(
     args: clientUtils.ClientBaseParams,
   ): Promise<clientUtils.Response<clientUtils.ResponseValue | clientUtils.ResponseValue[]>> {
+    const rateLimits = this.rateLimits.find(({ url }) => args.url.match(url))?.limits ?? this.defaultRateLimits
+    // If this is the initial request, don't wait
+    if (!_.isEqual(rateLimits, initRateLimits)
+      // If nothing is running, and we passed the rate limit reset time, don't wait
+      && !(rateLimits.currentlyRunning === 0 && rateLimits.rateLimitReset < Date.now())
```

Comment:
not sure I understand this one - why do we need to check that nothing is running?

also (and related), I think the negation is complicating things here - IMO this would be easier to understand as
```suggestion
      && (rateLimits.currentlyRunning !== 0 || rateLimits.rateLimitReset >= Date.now())
```
we might need some other handling for this, but I think recursing and sleeping every time there's a concurrent call is not ideal (if we could avoid recursing it might be ok since the `sleep` function has an early exit if the sleep time is non-positive)

Reply:
There is a buffer on the sleep to avoid extreme recursing, but i change it into a while loop and it should be equal
also removed the currentlyRunning === 0, it does seem like the other condition covers it